internal/cmd/io_binary.go

```diff
@@ -0,0 +1,267 @@
+	"sync"
+	"time"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/pkg/errors"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+)
+
+const (
+	binaryPipeFmt         = `\\.\pipe\binary-%s-%s`
+	binaryCmdWaitTimeout  = 10 * time.Second
+	binaryCmdStartTimeout = 10 * time.Second
+)
+
+// NewBinaryIO runs a custom binary process for pluggable shim logging driver.
+//
+// Container's IO will be redirected to the logging driver via named pipes, which are
+// passed as "CONTAINER_STDOUT", "CONTAINER_STDERR" environment variables. The logging
+// driver MUST dial a wait pipe passed via "CONTAINER_WAIT" environment variable AND CLOSE
+// it to indicate that it's ready to consume the IO. For customer's convenience container ID
+// and namespace are also passed via "CONTAINER_ID" and "CONTAINER_NAMESPACE".
+//
+// The path to the logging driver can be provided via a URL's host/path. Additional arguments
+// can be passed to the logger via URL query params
+func NewBinaryIO(ctx context.Context, id string, uri *url.URL) (_ UpstreamIO, err error) {
+	ns, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		ns = namespaces.Default
+	}
+
+	var sout, serr, w io.ReadWriteCloser
+
+	stdoutPipe := fmt.Sprintf(binaryPipeFmt, id, "stdout")
+	sout, err = openNPipe(stdoutPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	stderrPipe := fmt.Sprintf(binaryPipeFmt, id, "stderr")
+	serr, err = openNPipe(stderrPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	waitPipe := fmt.Sprintf(binaryPipeFmt, id, "wait")
+	w, err = openNPipe(waitPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	defer w.Close()
+
+	envs := []string{
+		"CONTAINER_ID=" + id,
+		"CONTAINER_NAMESPACE=" + ns,
+		"CONTAINER_STDOUT=" + stdoutPipe,
+		"CONTAINER_STDERR=" + stderrPipe,
+		"CONTAINER_WAIT=" + waitPipe,
+	}
+	cmd, err := newBinaryCmd(ctx, uri, envs)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+
+	errCh := make(chan error, 1)
+	go func() {
+		b := make([]byte, 1)
+		if _, err := w.Read(b); err != nil && err != io.EOF {
+			errCh <- err
+			return
+		}
+		errCh <- nil
+	}()
+
+	select {
+	case err = <-errCh:
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to start binary logger")
+		}
+	case <-time.After(binaryCmdStartTimeout):
+		return nil, errors.New("failed to start binary logger: timeout")
+	}
+
+	return &binaryIO{
+		cmd:    cmd,
+		stdout: stdoutPipe,
+		sout:   sout,
+		stderr: stderrPipe,
+		serr:   serr,
+	}, nil
+}
+
+func newBinaryCmd(ctx context.Context, uri *url.URL, envs []string) (*exec.Cmd, error) {
+	if uri.Host == "" && uri.Path == "" {
+		return nil, errors.New("no logging driver path provided")
+	}
+
+	var args []string
+	for k, vs := range uri.Query() {
+		args = append(args, k)
+		if len(vs) > 0 && vs[0] != "" {
+			args = append(args, vs[0])
+		}
+	}
+
+	execPath := uri.Path
+
+	if uri.Host != "" {
+		execPath = "/" + uri.Host + uri.Path
```

Comment:
What does this line do?

Reply:
we need an absolute path and this is done to treat `binary://path/to/file` and `binary:///path/to/file` as the same. I'll add a comment